### ISSUE TITLE
Possible bad interaction between KSPUpgradePipeline, B9PS and TweakScale (also happening on Revert to Editor at least)

### ISSUE THREAD

**Issue by Lisias** [net-lisias-ksp/TweakScale#168]
Fellow Kerbonaut [alphaprior](https://forum.kerbalspaceprogram.com/index.php?/profile/185480-alphaprior/) [reported a problem](https://forum.kerbalspaceprogram.com/index.php?/topic/179030-ksp-144-tweakscale-under-lisias-management-2445-2021-0120/&do=findComment&comment=3928212) involving TweakScale.

```
[ERR 03:01:50.064] Exception handling event onEditorShipModified in class ModuleB9PartInfo:System.NullReferenceException: Object reference not set to an instance of an o
  at TweakScale.TweakScale.IPartMassModifier.GetModuleMass (System.Single defaultMass, ModifierStagingSituation situation) [0x00010] in <6c196c229b7b407390d4d14556b33c91
  at Part.GetModuleMass (System.Single defaultMass, ModifierStagingSituation sit) [0x00046] in <2afc64dea36946459d4707808bdac511>:0
  at B9PartSwitch.ModuleB9PartInfo.UpdateFields () [0x00019] in <b85eabe067044bcb9aeabe863954a5ce>:0
  at B9PartSwitch.ModuleB9PartInfo.EditorShipModified (ShipConstruct construct) [0x00006] in <b85eabe067044bcb9aeabe863954a5ce>:0
  at EventData`1[T].Fire (T data) [0x000b0] in <2afc64dea36946459d4707808bdac511>:0

[EXC 03:01:50.067] NullReferenceException: Object reference not set to an instance of an object
    TweakScale.TweakScale.IPartMassModifier.GetModuleMass (System.Single defaultMass, ModifierStagingSituation situation) (at <6c196c229b7b407390d4d14556b33c91>:0)
    Part.GetModuleMass (System.Single defaultMass, ModifierStagingSituation sit) (at <2afc64dea36946459d4707808bdac511>:0)
    B9PartSwitch.ModuleB9PartInfo.UpdateFields () (at <b85eabe067044bcb9aeabe863954a5ce>:0)
    B9PartSwitch.ModuleB9PartInfo.EditorShipModified (ShipConstruct construct) (at <b85eabe067044bcb9aeabe863954a5ce>:0)
    EventData`1[T].Fire (T data) (at <2afc64dea36946459d4707808bdac511>:0)
    UnityEngine.DebugLogHandler:LogException(Exception, Object)
    ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
    UnityEngine.Debug:LogException(Exception)
    EventData`1:Fire(ShipConstruct)
    EditorLogic:SetBackup()
    EditorLogic:StartEditor(Boolean)
    EditorLogic:ShipToLoadSelected(ConfigNode, LoadType)
    KSP.UI.Screens.CraftBrowserDialog:onPipelineFinished(ConfigNode, CraftEntry, LoadType)
    KSP.UI.Screens.<>c__DisplayClass126_0:<pipeSelectedItem>b__0(ConfigNode)
    KSPUpgradePipeline:Process(ConfigNode, String, LoadContext, Callback`1, Callback`2)
    KSP.UI.Screens.CraftBrowserDialog:pipeSelectedItem(CraftEntry, LoadType)
    KSP.UI.Screens.CraftBrowserDialog:ConfirmLoadCraft()
    KSP.UI.Screens.CraftBrowserDialog:onButtonLoad()
    UnityEngine.EventSystems.EventSystem:Update()
[LOG 03:01:50.261] [Part truss-spinal-02] [ModuleB9PartSwitch 'meshSwitch'] Switched subtype to None
```

> The KSPUpgradePipeline:Process is involved on the mess! THIS may be trigger for the problem and the reason I can't reproduce the problem on my rig - my test beds are most of the time "clean", I usually don't have artefacts from previous KSP versions around - that it's when the KSPUpgradePipeline thingy is invoked...

[KSP.log.zip](https://github.com/net-lisias-ksp/TweakScale/files/6003021/KSP.log.zip)

TODO:

1. Reproduce the damned thing on a test bed
2. Check if the hypothesis above is valid
3. See how TweakScale can workaround the problem - a dummy PartBD on Awake perhaps?

**Comment by Lisias** [comment]
[AccidentalDisassembly](https://forum.kerbalspaceprogram.com/index.php?/profile/110936-accidentaldisassembly/) is getting [some heat](https://forum.kerbalspaceprogram.com/index.php?/topic/179030-ksp-144-tweakscale-under-lisias-management-2445-2021-0120/&do=findComment&comment=3928409) from B9PS and TweakScale too, and the thing is probably be related. So I will report my findings here.

Problem - B9PS and TweakScale are behaving flawless on my rig. :(

![screenshot51](https://user-images.githubusercontent.com/64334/108440911-ce4cc100-7232-11eb-9264-395bb282febe.png)

I already have evidence that we have a 3rd party borking something somewhere, and this bork is screwing up TweakScale somehow. I will try to tackle down the issue at hand (the Upgrade Pipeline thingy) and then see if this solves the other issues too.

**Comment by Lisias** [comment]
I'm an idiot. :) I forgot to **scale the damned part!!**.

Once the part is scaled, the Exceptions start to blow on the KSP.log:

```
[ERR 21:50:51.438] Exception handling event onEditorShipModified in class ModuleB9PartInfo:System.NullReferenceException
  at (wrapper managed-to-native) UnityEngine.Behaviour.get_enabled(UnityEngine.Behaviour)
  at TweakScale.TweakScale.get_IsScaled () [0x00000] in <d0c485183fcf4c76a03b2923576b7ed2>:0
  at TweakScale.TweakScale.IPartMassModifier.GetModuleMass (System.Single defaultMass, ModifierStagingSituation situation) [0x00000] in <d0c485183fcf4c76a03b2923576b7ed2
  at Part.GetModuleMass (System.Single defaultMass, ModifierStagingSituation sit) [0x00046] in <dcd149aee9c64162a3bb6c8525068846>:0
  at B9PartSwitch.ModuleB9PartInfo.UpdateFields () [0x00019] in <b85eabe067044bcb9aeabe863954a5ce>:0
  at B9PartSwitch.ModuleB9PartInfo.EditorShipModified (ShipConstruct construct) [0x00006] in <b85eabe067044bcb9aeabe863954a5ce>:0
  at EventData`1[T].Fire (T data) [0x000b0] in <dcd149aee9c64162a3bb6c8525068846>:0

[EXC 21:50:51.444] NullReferenceException
    TweakScale.TweakScale.get_IsScaled () (at <d0c485183fcf4c76a03b2923576b7ed2>:0)
    TweakScale.TweakScale.IPartMassModifier.GetModuleMass (System.Single defaultMass, ModifierStagingSituation situation) (at <d0c485183fcf4c76a03b2923576b7ed2>:0)
    Part.GetModuleMass (System.Single defaultMass, ModifierStagingSituation sit) (at <dcd149aee9c64162a3bb6c8525068846>:0)
    B9PartSwitch.ModuleB9PartInfo.UpdateFields () (at <b85eabe067044bcb9aeabe863954a5ce>:0)
    B9PartSwitch.ModuleB9PartInfo.EditorShipModified (ShipConstruct construct) (at <b85eabe067044bcb9aeabe863954a5ce>:0)
    EventData`1[T].Fire (T data) (at <dcd149aee9c64162a3bb6c8525068846>:0)
    UnityEngine.DebugLogHandler:LogException(Exception, Object)
    KSPe.Util.Log.UnityLogDecorator:UnityEngine.ILogHandler.LogException(Exception, Object)
    ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
    UnityEngine.Debug:LogException(Exception)
    EventData`1:Fire(ShipConstruct)
    EditorLogic:SetBackup()
    EditorLogic:StartEditor(Boolean)
    <Start>d__133:MoveNext()
    UnityEngine.SetupCoroutine:InvokeMoveNext(IEnumerator, IntPtr)
[LOG 21:50:51.445] [VesselMover] INFO: [Move Launch Controller]: Hiding Flight GUI
```

I'm reasonably sure that this is also the root cause for the Upgrade Pipeline bork,

**Comment by Lisias** [comment]
Now I'm confused. I had assumed I had borked something on the Awake event while cloning, but I was wrong. I did it right.

So back to basics, this is what I'm getting on my KSP.log:

```
[ERR 11:32:28.794] Exception handling event onEditorShipModified in class ModuleB9PartInfo:System.NullReferenceException
  at (wrapper managed-to-native) UnityEngine.Behaviour.get_enabled(UnityEngine.Behaviour)
  at TweakScale.TweakScale.get_IsScaled () [0x00000] in <7a868806c74a437a9c825ec69ac6d367>:0
  at TweakScale.TweakScale.IPartMassModifier.GetModuleMass (System.Single defaultMass, ModifierStagingSituation situation) [0x00000] in <7a868806c74a437a9c825ec69ac6d367>:0
  at Part.GetModuleMass (System.Single defaultMass, ModifierStagingSituation sit) [0x00046] in <dcd149aee9c64162a3bb6c8525068846>:0
  at B9PartSwitch.ModuleB9PartInfo.UpdateFields () [0x00019] in <b85eabe067044bcb9aeabe863954a5ce>:0
  at B9PartSwitch.ModuleB9PartInfo.EditorShipModified (ShipConstruct construct) [0x00006] in <b85eabe067044bcb9aeabe863954a5ce>:0
  at EventData`1[T].Fire (T data) [0x000b0] in <dcd149aee9c64162a3bb6c8525068846>:0
```

1. The B9PS EditorShipModified is fired
2. Something on B9PS start to update its fields, what leads to
3. Part.GetModuleMass being called for each module
4. So is unsurprisngly that TweakScale.TweakScale.IPartMassModifier.GetModuleMass is called eventually.
5. That so calls TweakScale.TweakScale.get_IsScaled to check if the part is scaled (when it is not, it returns 0).

So the problem is the IsScaled property, right?

Well, nope!

```
public bool IsScaled => this.enabled && (Math.Abs(currentScale / defaultScale - 1f) > 1e-5f);
```
https://github.com/net-lisias-ksp/TweakScale/blob/b45cca5a2dabf1c6d5227c9819e2242c9259c08c/Source/Scale/Scale.cs#L135

O `this.enabled` is inherited from MonoBehaviour, and `currentScale` and `defaultScale` are float class members .

`this` obviously is not null, so obviously the problem is coming from Unity itself, see the `UnityEngine.Behaviour.get_enabled` on the KSP.log!!!!

**Comment by Lisias** [comment]
Reopening, as I clicked on the wrong button!! :)

**Comment by Lisias** [comment]
Hummm... Nope. Github has a new (unwanted) feature, it closes issues by itself if you use some keywords on commits. :(

**Comment by Lisias** [comment]
Anyway. I **FIXED** the problem! :)

B9PS was failing from remove itself from the OnEditorShipModified game event when it was being destroyed (what happens when you revert to the editor), and so dead objects were being fired on the Editor. Such dead objets had references for more dead objects from others Add'Ons (as TweakScale), and this appear to be the reason the bork had happened inside Unity - the Scale object that the dead B9PS object had fired was on the process had already "killed" itself (I handle these events), and also means that Unity had already destroyed its own control info on that Scale Object!

**Comment by Lisias** [comment]
Pull request to the upstream were applied:

https://github.com/blowfishpro/B9PartSwitch/pull/207

**Comment by Lisias** [comment]
(sigh). I need to locate and deactivate that kraken damned workflow.

I intent to keep this open until the new release of B9 Part Switch is on the wild.

**Comment by Lisias** [comment]
Pull request was accepted and merged.

https://github.com/blowfishpro/B9PartSwitch/pull/207

Don't have a clue about when the new release with the fix will be published.

**Comment by Lisias** [comment]
Fix was released on B9PS 2.18.0.

https://github.com/blowfishpro/B9PartSwitch/releases/tag/v2.18.0

So I'm (finally) closing this.
